Build failure: vmware-workstation

### Steps To Reproduce

Steps to reproduce the behavior:
1. build *vmware-workstation*

### Build log

```
building Nix...
building the system configuration...
building '/nix/store/vh6yjal2lyp9va5k5p5shv073jzsgay3-VMware-Workstation-Full-17.5.1-23298084.x86_64.bundle.drv'...
building '/nix/store/4slcz39jjfqb73rjqc2kbqqb9i6nppjy-gdbm-1.8.3.drv'...
building '/nix/store/gqwx8nkfw2m5xw92y6k3s4a3a2iqzrzn-vmware-unpack-env-usr-target.drv'...
Running phase: unpackPhase
unpacking source archive /nix/store/hh149gw57kdvafpqjfq3869x4zkkgdr7-gdbm-1.8.3.tar.gz
source root is gdbm-1.8.3
setting SOURCE_DATE_EPOCH to timestamp 1034719556 of file gdbm-1.8.3/gdbm.info
Running phase: patchPhase
Running phase: updateAutotoolsGnuConfigScriptsPhase
Updating Autotools / GNU config script to a newer upstream version: ./config.sub
Updating Autotools / GNU config script to a newer upstream version: ./config.guess
Running phase: configurePhase
fixing libtool script ./ltmain.sh
patching script interpreter paths in ./configure
./configure: interpreter directive changed from "#! /bin/sh" to "/nix/store/h3bhzvz9ipglcybbcvkxvm4vg9lwvqg4-bash-5.2p26/bin/sh"
configure flags: --disable-static --prefix=/nix/store/2gyp500pg8byyrbd9n4j6ijpbgb9ry4z-gdbm-1.8.3 --enable-libgdbm-compat
warning: collision between `/nix/store/nkdgysx90pln2l562l7lsk021hbxjdg8-glibc-2.39-31-bin/bin/ldconfig' and `/nix/store/d130xwm9365hfdzpz7x5kq9is9q6l4rm-ldconfig/bin/ldconfig'
warning: collision between `/nix/store/apab5i73dqa09wx0q27b6fbhd1r18ihl-glibc-2.39-31/lib/locale/locale-archive' and `/nix/store/j2h103h7b0xg7pc5ih3m1df0j3z71c6v-glibc-locales-2.39-31/lib/locale/locale-archive'
checking for gcc... gcc
created 1125 symlinks in user environment
building '/nix/store/jyymlbz15czjll8hac5qddil8afar8rf-vmware-unpack-env-fhs.drv'...
checking for C compiler default output... a.out
checking whether the C compiler works... yes
checking whether we are cross compiling... no
checking for suffix of executables... 
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
/nix/store/4hfpf6kn1fj2r1dxpn8dz7zqdnm3j12l-vmware-unpack-env-fhs /build
/nix/store/4hfpf6kn1fj2r1dxpn8dz7zqdnm3j12l-vmware-unpack-env-fhs/usr /nix/store/4hfpf6kn1fj2r1dxpn8dz7zqdnm3j12l-vmware-unpack-env-fhs /build
/nix/store/4hfpf6kn1fj2r1dxpn8dz7zqdnm3j12l-vmware-unpack-env-fhs /build
building '/nix/store/pl6gzr7sjfznn331rkp6iqs439nxav3s-vmware-unpack-env-bwrap.drv'...

trying https://download3.vmware.com/software/WKST-1751-LX/VMware-Workstation-Full-17.5.1-23298084.x86_64.bundle
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
checking for gcc option to accept ANSI C... none needed
checking how to run the C preprocessor... gcc -E
checking for a BSD-compatible install... /nix/store/hq8765g3p1i7qbargnqli5mn0jpsdbfl-coreutils-9.5/bin/install -c
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-pc-linux-gnu
checking for ld used by GCC... ld
checking if the linker (ld) is GNU ld... yes
checking for ld option to reload object files... -r
checking for BSD-compatible nm... nm
checking whether ln -s works... yes
checking how to recognise dependant libraries... (cached) pass_all
building '/nix/store/5l3rqri02lk0vlsshzxavdsd240zcm30-vmware-unpack-env.drv'...
checking command to parse nm output... ok
checking for egrep... grep -E
  0    15    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (56) The requested URL returned error: 522
error: cannot download VMware-Workstation-Full-17.5.1-23298084.x86_64.bundle from any mirror
checking for ANSI C header files... yes
error: builder for '/nix/store/vh6yjal2lyp9va5k5p5shv073jzsgay3-VMware-Workstation-Full-17.5.1-23298084.x86_64.bundle.drv' failed with exit code 1
error:
       … while calling the 'head' builtin

         at /nix/var/nix/profiles/per-user/root/channels/nixos/lib/attrsets.nix:1575:11:

         1574|         || pred here (elemAt values 1) (head values) then
         1575|           head values
             |           ^
         1576|         else

       … while evaluating the attribute 'value'

         at /nix/var/nix/profiles/per-user/root/channels/nixos/lib/modules.nix:809:9:

          808|     in warnDeprecation opt //
          809|       { value = builtins.addErrorContext "while evaluating the option `${showOption loc}':" value;
             |         ^
          810|         inherit (res.defsFinal') highestPrio;

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: 1 dependencies of derivation '/nix/store/6zl9d3wlrw7s6p63l9agnangra8yfqdm-vmware-workstation-17.5.1.drv' failed to build
```

### Additional context

The package cannot download. 522 error. It is similarly impossible to install a previous version. 

### Notify maintainers

<!--
home@chrisaw.com
vifino@tty.sh
-->

I have written an overlay and adapted it to my configuration. The overwrite works successfully (I just used your github repo temporarily). 

```

    vmware-workstation = prev.vmware-workstation.overrideAttrs (old: {
	src = final.fetchurl {
	url = "https://github.com/kulczwoj/vmware-workstation/releases/download/v${builtins.head (final.lib.splitString "." old.version)}.0/VMware-Workstation-Full-${old.version}-${old.build}.x86_64.bundle";
	sha256 = "sha256-qmC3zvKoes77z3x6UkLHsJ17kQrL1a/rxe9mF+UMdJY=";
	};
	unpackPhase = let
	vmware-unpack-env = final.buildFHSEnv {
	name = "vmware-unpack-env";
	targetPkgs = pkgs: [ final.zlib ];
	};
	in ''
	${vmware-unpack-env}/bin/vmware-unpack-env -c "sh ${old.src} --extract unpacked"
	'';
	});
```



Unfortunately I get a build error due to the `linux-6.9-dev/lib/modules.`

```
rror: builder for '/nix/store/ny4hpr1736dpvfqkhcbh4d85jrbsyfdj-vmware-modules-workstation-17.5.1-unstable-2024-01-12-6.9.drv' failed with exit code 2;
       last 10 log lines:
       >   BTF [M] /build/source/vmmon-only/vmmon.ko
       > Skipping BTF generation for /build/source/vmmon-only/vmmon.ko due to unavailability of vmlinux
       > make[2]: Leaving directory '/nix/store/7j5krfzkiw31vkkv1jy13lpfzkiryn5s-linux-6.9-dev/lib/modules/6.9.0/build'
       > make -C $PWD SRCROOT=$PWD/. \
       >   MODULEBUILDDIR= postbuild
       > make[2]: Entering directory '/build/source/vmmon-only'
       > make[2]: 'postbuild' is up to date.
       > make[2]: Leaving directory '/build/source/vmmon-only'
       > cp -f vmmon.ko ./../vmmon.o
       > make[1]: Leaving directory '/build/source/vmmon-only'
       For full logs, run 'nix log /nix/store/ny4hpr1736dpvfqkhcbh4d85jrbsyfdj-vmware-modules-workstation-17.5.1-unstable-2024-01-12-6.9.drv'.
error: 1 dependencies of derivation '/nix/store/gfcbzcw13wzc9jgnbhz4ny1dwhjv3k2x-linux-6.9-modules.drv' failed to build
error: 1 dependencies of derivation '/nix/store/mm19xvlpn1vwvga9n7xxwcwsvmiicxdi-system-path.drv' failed to build
error: 1 dependencies of derivation '/nix/store/1hm0ymnbmsbwrf1vj2bj1j8a1axi9dgk-nixos-system-sfs-24.05.20240516.bf446f0.drv' failed to build
```

```
In file included from /nix/store/7j5krfzkiw31vkkv1jy13lpfzkiryn5s-linux-6.9-dev/lib/modules/6.9.0/source/include/linux/s>
                 from /nix/store/7j5krfzkiw31vkkv1jy13lpfzkiryn5s-linux-6.9-dev/lib/modules/6.9.0/source/include/linux/s>
                 from /build/source/vmnet-only/bridge.c:25:
/build/source/vmnet-only/bridge.c: In function 'VNetBridgeReceiveFromVNet':
/build/source/vmnet-only/vmnetInt.h:44:39: error: 'dev_base_lock' undeclared (first use in this function); did you mean >
   44 | #define dev_lock_list()    read_lock(&dev_base_lock)
      |                                       ^~~~~~~~~~~~~
/nix/store/7j5krfzkiw31vkkv1jy13lpfzkiryn5s-linux-6.9-dev/lib/modules/6.9.0/source/include/linux/rwlock.h:56:48: note: i>
   56 | #define read_lock(lock)         _raw_read_lock(lock)
      |                                                ^~~~
/build/source/vmnet-only/bridge.c:587:4: note: in expansion of macro 'dev_lock_list'
  587 |    dev_lock_list();
      |    ^~~~~~~~~~~~~
```

I have updated my kernel to `6.9.0.` 
Which kernel / vmware combination are you currently using?

> I have written an overlay and adapted it to my configuration. The overwrite works successfully (I just used your github repo temporarily).
> 
> ```
> 
>     vmware-workstation = prev.vmware-workstation.overrideAttrs (old: {
> 	src = final.fetchurl {
> 	url = "https://github.com/kulczwoj/vmware-workstation/releases/download/v${builtins.head (final.lib.splitString "." old.version)}.0/VMware-Workstation-Full-${old.version}-${old.build}.x86_64.bundle";
> 	sha256 = "sha256-qmC3zvKoes77z3x6UkLHsJ17kQrL1a/rxe9mF+UMdJY=";
> 	};
> 	unpackPhase = let
> 	vmware-unpack-env = final.buildFHSEnv {
> 	name = "vmware-unpack-env";
> 	targetPkgs = pkgs: [ final.zlib ];
> 	};
> 	in ''
> 	${vmware-unpack-env}/bin/vmware-unpack-env -c "sh ${old.src} --extract unpacked"
> 	'';
> 	});
> ```
> 
> Unfortunately I get a build error due to the `linux-6.9-dev/lib/modules.`
> 
> ```
> rror: builder for '/nix/store/ny4hpr1736dpvfqkhcbh4d85jrbsyfdj-vmware-modules-workstation-17.5.1-unstable-2024-01-12-6.9.drv' failed with exit code 2;
>        last 10 log lines:
>        >   BTF [M] /build/source/vmmon-only/vmmon.ko
>        > Skipping BTF generation for /build/source/vmmon-only/vmmon.ko due to unavailability of vmlinux
>        > make[2]: Leaving directory '/nix/store/7j5krfzkiw31vkkv1jy13lpfzkiryn5s-linux-6.9-dev/lib/modules/6.9.0/build'
>        > make -C $PWD SRCROOT=$PWD/. \
>        >   MODULEBUILDDIR= postbuild
>        > make[2]: Entering directory '/build/source/vmmon-only'
>        > make[2]: 'postbuild' is up to date.
>        > make[2]: Leaving directory '/build/source/vmmon-only'
>        > cp -f vmmon.ko ./../vmmon.o
>        > make[1]: Leaving directory '/build/source/vmmon-only'
>        For full logs, run 'nix log /nix/store/ny4hpr1736dpvfqkhcbh4d85jrbsyfdj-vmware-modules-workstation-17.5.1-unstable-2024-01-12-6.9.drv'.
> error: 1 dependencies of derivation '/nix/store/gfcbzcw13wzc9jgnbhz4ny1dwhjv3k2x-linux-6.9-modules.drv' failed to build
> error: 1 dependencies of derivation '/nix/store/mm19xvlpn1vwvga9n7xxwcwsvmiicxdi-system-path.drv' failed to build
> error: 1 dependencies of derivation '/nix/store/1hm0ymnbmsbwrf1vj2bj1j8a1axi9dgk-nixos-system-sfs-24.05.20240516.bf446f0.drv' failed to build
> ```
> 
> ```
> In file included from /nix/store/7j5krfzkiw31vkkv1jy13lpfzkiryn5s-linux-6.9-dev/lib/modules/6.9.0/source/include/linux/s>
>                  from /nix/store/7j5krfzkiw31vkkv1jy13lpfzkiryn5s-linux-6.9-dev/lib/modules/6.9.0/source/include/linux/s>
>                  from /build/source/vmnet-only/bridge.c:25:
> /build/source/vmnet-only/bridge.c: In function 'VNetBridgeReceiveFromVNet':
> /build/source/vmnet-only/vmnetInt.h:44:39: error: 'dev_base_lock' undeclared (first use in this function); did you mean >
>    44 | #define dev_lock_list()    read_lock(&dev_base_lock)
>       |                                       ^~~~~~~~~~~~~
> /nix/store/7j5krfzkiw31vkkv1jy13lpfzkiryn5s-linux-6.9-dev/lib/modules/6.9.0/source/include/linux/rwlock.h:56:48: note: i>
>    56 | #define read_lock(lock)         _raw_read_lock(lock)
>       |                                                ^~~~
> /build/source/vmnet-only/bridge.c:587:4: note: in expansion of macro 'dev_lock_list'
>   587 |    dev_lock_list();
>       |    ^~~~~~~~~~~~~
> ```
> 
> I have updated my kernel to `6.9.0.` Which kernel / vmware combination are you currently using?

I'm using 6.6.30. I used to use latest, but switched to LTS because of VMware kernel modules and constant problems with them.

--- a/pkgs/applications/virtualization/vmware-workstation/default.nix
+++ b/pkgs/applications/virtualization/vmware-workstation/default.nix
@@ -1,14 +1,15 @@
 { stdenv
 , buildFHSEnv
 , fetchurl
+, fetchzip
 , lib
 , zlib
 , gdbm
-, bzip2
 , libxslt
 , libxml2
 , libuuid
 , readline
+, readline70
 , xz
 , cups
 , glibc
@@ -25,22 +26,50 @@
 , python3
 , autoPatchelfHook
 , makeWrapper
-, sqlite
-, enableInstaller ? false
-, enableMacOSGuests ? false, fetchFromGitHub, gnutar, unzip
+, symlinkJoin
+, enableInstaller ? false, bzip2, sqlite
+, enableMacOSGuests ? false, fetchFromGitHub, unzip
+, enableGuestTools ? true,
 }:
 
 let
+  # base - versions
+  version = "17.5.2";
+  build = "23775571";
+  baseUrl = "https://softwareupdate.vmware.com/cds/vmw-desktop/ws/${version}/${build}/linux";
+
+  # tools - versions
+  toolsVersion = "12.4.0";
+  toolsBuild = "23259341";
+
   # macOS - versions
-  fusionVersion = "13.5.1";
-  fusionBuild = "23298085";
+  fusionVersion = "13.5.2";
+  fusionBuild = "23775688";
   unlockerVersion = "3.0.5";
 
-  # macOS - ISOs
-  darwinIsoSrc = fetchurl {
-    url = "https://softwareupdate.vmware.com/cds/vmw-desktop/fusion/${fusionVersion}/${fusionBuild}/universal/core/com.vmware.fusion.zip.tar";
-    sha256 = "sha256-bn6hoicby2YVj1pZTBzBhabNhKefzVQTm5vIrdTO2K4=";
+  guestToolsSrc =
+  let
+    fetchComponent = (system: hash: fetchzip {
+      inherit hash;
+      url = "${baseUrl}/packages/vmware-tools-${system}-${toolsVersion}-${toolsBuild}.x86_64.component.tar";
+      stripRoot = false;
+    } + "/vmware-tools-${system}-${toolsVersion}-${toolsBuild}.x86_64.component");
+  in lib.mapAttrsToList fetchComponent {
+      linux = "sha256-vT08mR6cCXZjiQgb9jy+MaqYzS0hFbNUM7xGAHIJ8Ao=";
+      linuxPreGlibc25 = "sha256-BodN1lxuhxyLlxIQSlVhGKItJ10VPlti/sEyxcRF2SA=";
+      netware = "sha256-o/S4wAYLR782Fn20fTQ871+rzsa1twnAxb9laV16XIk=";
+      solaris = "sha256-3LdFoI4TD5zxlohDGR3DRGbF6jwDZAoSMEpHWU4vSGU=";
+      winPre2k = "sha256-+QcvWfY3aCDxUwAfSuj7Wf9sxIO+ztWBrRolMim8Dfw=";
+      winPreVista = "sha256-3NgO/GdRFTpKNo45TMet0msjzxduuoF4nVLtnOUTHUA=";
+      windows = "sha256-2F7UPjNvtibmWAJxpB8IOnol12aMOGMy+403WeCTXw8=";
   };
+
+  # macOS - ISOs
+  darwinIsoSrc = fetchzip {
+    url = "https://softwareupdate.vmware.com/cds/vmw-desktop/fusion/${fusionVersion}/${fusionBuild}/universal/core/com.vmware.fusion.zip.tar";
+    sha256 = "sha256-DDLRWAVRI3ZeXV5bUXWwput9mEC1qsJUsjojI0CJYMI=";
+    stripRoot = false;
+  } + "/com.vmware.fusion.zip";
 
   # macOS - Unlocker
   unlockerSrc = fetchFromGitHub {
@@ -68,11 +97,18 @@ let
     name = "vmware-unpack-env";
     targetPkgs = pkgs: [ zlib ];
   };
+
+  readline70_compat63 = symlinkJoin {
+    name = "readline70_compat63";
+    paths = [ readline70 ];
+    postBuild = ''
+      ln -s $out/lib/libreadline.so $out/lib/libreadline.so.6
+    '';
+  };
 in
 stdenv.mkDerivation rec {
   pname = "vmware-workstation";
-  version = "17.5.1";
-  build = "23298084";
+  inherit version build;
 
   buildInputs = [
     libxslt
@@ -96,21 +132,24 @@ stdenv.mkDerivation rec {
   ];
 
   nativeBuildInputs = [ python3 vmware-unpack-env autoPatchelfHook makeWrapper ]
-    ++ lib.optionals enableInstaller [ sqlite bzip2 ]
-    ++ lib.optionals enableMacOSGuests [ gnutar unzip ];
+    ++ lib.optionals enableInstaller [ bzip2 sqlite readline70_compat63 ]
+    ++ lib.optionals enableMacOSGuests [ unzip ];
 
-  src = fetchurl {
-    url = "https://download3.vmware.com/software/WKST-${builtins.replaceStrings ["."] [""] version}-LX/VMware-Workstation-Full-${version}-${build}.x86_64.bundle";
-    sha256 = "sha256-qmC3zvKoes77z3x6UkLHsJ17kQrL1a/rxe9mF+UMdJY=";
-  };
+  src = fetchzip {
+    url = "${baseUrl}/core/VMware-Workstation-${version}-${build}.x86_64.bundle.tar";
+    sha256 = "sha256-5PZZpXN/V687TXjqeTm8MEays4/QTf02jVfdpi9C7GI=";
+    stripRoot = false;
+  } + "/VMware-Workstation-${version}-${build}.x86_64.bundle";
 
-  unpackPhase = ''
-    ${vmware-unpack-env}/bin/vmware-unpack-env -c "sh ${src} --extract unpacked"
+  unpackPhase = let
+    guestTools = lib.optionalString enableGuestTools (lib.concatMapStringsSep " " (src: "--install-component ${src}") guestToolsSrc);
+  in
+  ''
+    ${vmware-unpack-env}/bin/vmware-unpack-env -c "sh ${src} ${guestTools} --extract unpacked"
 
     ${lib.optionalString enableMacOSGuests ''
       mkdir -p fusion/
-      tar -xvpf "${darwinIsoSrc}" -C fusion/
-      unzip "fusion/com.vmware.fusion.zip" \
+      unzip "${darwinIsoSrc}" \
         "payload/VMware Fusion.app/Contents/Library/isoimages/x86_x64/darwin.iso" \
         "payload/VMware Fusion.app/Contents/Library/isoimages/x86_x64/darwinPre15.iso" \
         -d fusion/
@@ -173,7 +212,7 @@ stdenv.mkDerivation rec {
         component_version=$(cat unpacked/$component/manifest.xml | grep -oPm1 "(?<=<version>)[^<]+")
         component_core_id=$([ "$component" == "vmware-installer" ] && echo "-1" || echo "1")
         type=$([ "$component" == "vmware-workstation" ] && echo "0" || echo "1")
-        sqlite3 "$database_filename" "INSERT INTO components(name,version,buildNumber,component_core_id,longName,description,type) VALUES(\"$component\",\"$component_version\",\"${build}\",$component_core_id,\"$component\",\"$component\",$type);"
+        sqlite3 "$database_filename" "INSERT INTO components(name,version,buildNumber,component_core_id,longName,description,type) VALUES('$component','$component_version',${build},$component_core_id,'$component','$component',$type);"
         mkdir -p $out/etc/vmware-installer/components/$component
         cp -r $folder/* $out/etc/vmware-installer/components/$component
       done
@@ -255,9 +294,10 @@ stdenv.mkDerivation rec {
     unpacked="unpacked/vmware-network-editor"
     cp -r $unpacked/lib $out/lib/vmware/
 
-    ## VMware Tools
-    echo "Installing VMware Tools"
     mkdir -p $out/lib/vmware/isoimages/
+
+    ${lib.optionalString enableGuestTools ''
+    echo "Installing VMware Tools"
     cp unpacked/vmware-tools-linux/linux.iso \
        unpacked/vmware-tools-linuxPreGlibc25/linuxPreGlibc25.iso \
        unpacked/vmware-tools-netware/netware.iso \
@@ -266,6 +306,7 @@ stdenv.mkDerivation rec {
        unpacked/vmware-tools-winPreVista/winPreVista.iso \
        unpacked/vmware-tools-windows/windows.iso \
        $out/lib/vmware/isoimages/
+    ''}
 
     ${lib.optionalString enableMacOSGuests ''
       echo "Installing VMWare Tools for MacOS"

